### PR TITLE
[llvm][Support] ListeningSocket::accept returns operation_canceled if FD is set to -1

### PR DESCRIPTION
**Pull request by cpsughrue**
If `::poll` returns and `FD` equals -1, then `ListeningSocket::shutdown` has been called. So, regardless of any other information that could be gleaned from `FDs.revents` or `PollStatus`, it is appropriate to return `std::errc::operation_canceled`. `ListeningSocket::shutdown` copies `FD`'s value to `ObservedFD` then sets `FD` to -1 before canceling `::poll` by calling `::close(ObservedFD)` and writing to the pipe.

### CODE REVIEW

File: llvm/lib/Support/raw_socket_stream.cpp [hunk 1]
```diff
@@ -200,35 +200,40 @@ ListeningSocket::accept(std::chrono::milliseconds Timeout) {
   while (PollStatus == -1 && (Timeout.count() == -1 || ElapsedTime < Timeout)) {
     if (Timeout.count() != -1)
       RemainingTime -= ElapsedTime.count();
-
     auto Start = std::chrono::steady_clock::now();
+
```
Comment (Bigcheese) [inline llvm/lib/Support/raw_socket_stream.cpp:204]: Extra whitespace changes.

File: llvm/lib/Support/raw_socket_stream.cpp [hunk 2]
```diff
@@ -200,35 +200,40 @@ ListeningSocket::accept(std::chrono::milliseconds Timeout) {
...
 #ifdef _WIN32
     PollStatus = WSAPoll(FDs, 2, RemainingTime);
-    if (PollStatus == SOCKET_ERROR) {
 #else
     PollStatus = ::poll(FDs, 2, RemainingTime);
+#endif
+    // If FD equals -1 then ListeningSocket::shutdown has been called and it is
+    // appropriate to return operation_canceled. ListeningSocket::shutdown
+    // copies FD's value to ObservedFD then sets FD to -1 before canceling
+    // ::poll by calling close on ObservedFD and writing to the pipe.
```
Comment (Bigcheese) [inline llvm/lib/Support/raw_socket_stream.cpp:213]: Don't really need the second sentence here, people can go look at the implementation to see why the first sentence is true.

**Review by Bigcheese — COMMENTED** [review]
This looks correct with some minor comments, but it would be good to figure out why the existing shutdown test already returned `std::errc::operation_canceled`, but the module build daemon was seeing `std::errc::bad_file_descriptor` on shutdown.

**Comment by cpsughrue** [comment]
> This looks correct with some minor comments, but it would be good to figure out why the existing shutdown test already returned `std::errc::operation_canceled`, but the module build daemon was seeing `std::errc::bad_file_descriptor` on shutdown.

Still trying to figure out the root cause but the jist of the issue is that the below code does not fail
```
  llvm::handleAllErrors(std::move(Err), [&](const llvm::StringError &SE) {
    ASSERT_EQ(0,1);
  });
```
The raw_socket_stream FILE_DESCRIPTOR_CLOSED test should fail but wrapping a gtest macro in llvm::handleAllErrors seems to "silence" it.

**Comment by cpsughrue** [comment]
Another important note. The code
```
  ASSERT_THAT_EXPECTED(MaybeServer, Failed());
  llvm::Error Err = MaybeServer.takeError();
  llvm::handleAllErrors(std::move(Err), [&](const llvm::StringError &SE) {
    std::error_code EC = SE.convertToErrorCode();
    ASSERT_EQ(EC, std::errc::timed_out);
  });
```
results in two instances of `MaybeServer.takeError()` which is problematic and was also causing issues. Comments in `Error.h` explain that after calling `takeError`, `Expected<T>` is in an indeterminate state that can only be safely destructed. The hidden `MaybeServer.takeError()` takes place during the evaluation of ASSERT_THAT_EXPECTED.

**Review by Bigcheese — APPROVED** [review]
lgtm

**Pull request merged**